application/models/Permission.php

```diff
@@ -703,11 +706,20 @@ public function getButtons(): string
      */
     public function hasSurveyPermission($iSurveyID, $sPermission, $sCRUD = 'read', $iUserID = null)
     {
-        $oSurvey = Survey::Model()->findByPk($iSurveyID);
-        if (!$oSurvey) {
-            return false;
+        if (isset(self::$aCachedSurveyPermissions[$iSurveyID][$sPermission][$sCRUD][$iUserID])) {
+            return self::$aCachedSurveyPermissions[$iSurveyID][$sPermission][$sCRUD][$iUserID];
+        }
+        if (!isset(self::$aCachedSurveyPermissions[$iSurveyID])) {
+            self::$aCachedSurveyPermissions[$iSurveyID] = [];
         }
-        return $oSurvey->hasPermission($sPermission, $sCRUD, $iUserID);
```

Comment:
Already a partial static variables

https://github.com/LimeSurvey/LimeSurvey/blob/c10bac29fff26a55257a5b3d8997c18672e40eee/application/models/Permission.php#L619

Reply:
Maybe we can move add

```
if (isset($aPermissionStatic[$iEntityID][$sEntityName][$iUserID][$sPermission][$sCRUD])) {
return $aPermissionStatic[$iEntityID][$sEntityName][$iUserID][$sPermission][$sCRUD];
}
```

on top of https://github.com/LimeSurvey/LimeSurvey/blob/c10bac29fff26a55257a5b3d8997c18672e40eee/application/models/Permission.php#L548

(and add some set elsewhere https://github.com/LimeSurvey/LimeSurvey/blob/c10bac29fff26a55257a5b3d8997c18672e40eee/application/models/Permission.php#L558 https://github.com/LimeSurvey/LimeSurvey/blob/c10bac29fff26a55257a5b3d8997c18672e40eee/application/models/Permission.php#L566 https://github.com/LimeSurvey/LimeSurvey/blob/c10bac29fff26a55257a5b3d8997c18672e40eee/application/models/Permission.php#L571 https://github.com/LimeSurvey/LimeSurvey/blob/c10bac29fff26a55257a5b3d8997c18672e40eee/application/models/Permission.php#L595

Reply:
@Shnoulle this seems to be a good idea, worth checking, to apply the idea to permission checking in general. But, if costly things happen between survey permission checking and the generic permission checking, then this change might make sense as well. It's worth investigating.

Reply:
My opinion : currently : aPermissionStatic is badly used : we really need to use it before.

And it improve whol epermission system :)

Reply:
@Shnoulle I agree with your opinion. Yet, my point is that there are not one, but two issues to tackle with. First, we have the issue I tackled with here, regarding survey permissions, which has some benefit as long as it's not harmful (I will need to find it out, so if there is any harm, I will remove the caching system from oSurvey permission check) and the other topic is your suggestion for fixing the cache at the generic permission check.

I fully agree with you on the second and I will try to find out whether the first causes harm. If not, then it's a good change. If yes, then we will need to remove it. I will nevertheless look into prospects of generic permission checking.